File: points.rb
class Point
    attr_accessor :x, :y

    def initialize x, y
        @x = x.to_f
        @y = y.to_f
    end

    def self.gradient p1, p2
        (p2.y - p1.y) / (p2.x - p1.x)
    end

    def to_s
        "(#{@x.to_i},#{@y.to_i})"
    end
end

def discretePoints p1, p2, m
    (p1.x.to_i..p2.x.to_i).map { |x| Point.new(x, (m * (x.to_f - p2.x) + p2.y).round) }
end

point_1 = Point.new(15,3)
point_2 = Point.new(60,20)
gradient = Point.gradient point_1, point_2

puts discretePoints point_1, point_2, gradient

File: points.hs
discretePoints :: Int -> Int -> Int -> Int -> Double -> [(Int,Int)]
discretePoints x1 y1 x2 y2 m =
    [(i,j) | i <- [x1..x2],
             j <- [y1..y2],
             fromIntegral (j - y1) >= m * fromIntegral (i - x1) - 0.5,
             fromIntegral (j - y1) <= m * fromIntegral (i - x1) + 0.5]

type Point = (Int, Int)

discretePoints' :: Point -> Point -> Double -> [Point]
discretePoints' (x1,y1) (x2,y2) m =  [(x, f x) | x <- [x1..x2]]
  where
    f x = round (fromIntegral (x - x1) * m) + y1